parity/metrics.rs

```diff
@@ -0,0 +1,110 @@
+use std::{
+    time::Instant,
+    sync::{Mutex, Arc}
+};
+
+use crate::{
+    rpc,
+    rpc_apis
+};
+
+use hyper::{
+    Method, Body, Request, Response, Server, StatusCode,
+    service::{make_service_fn, service_fn}
+};
+
+use stats::{
+	PrometheusMetrics,
+	prometheus_gauge,
+	prometheus::{self, Encoder}
+};
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct MetricsConfiguration {
+	/// Are metrics enabled (default is false)?
+	pub enabled: bool,
+	/// The IP of the network interface used (default is 127.0.0.1).
+	pub interface: String,
+	/// The network port (default is 3000).
+	pub port: u16,
+}
+
+impl Default for MetricsConfiguration {
+	fn default() -> Self {
+		MetricsConfiguration {
+			enabled: false,
+			interface: "127.0.0.1".into(),
+			port: 3000,
+		}
+	}
+}
+
+struct State {
+	rpc_apis: Arc<rpc_apis::FullDependencies>,
+}
+
+fn handle_request(req: Request<Body>, state: &Arc<Mutex<State>>) -> Response<Body> {
+    let (parts, _body) = req.into_parts();
+    match (parts.method, parts.uri.path()) {
+        (Method::GET, "/metrics") => {
+            let start = Instant::now();
+            let mut reg = prometheus::Registry::new();
+
+            let state = state.lock().unwrap();
+
+            state.rpc_apis.client.prometheus_metrics(&mut reg);
```

Comment:
Please put the whole registry population logic into a future, spawn it with `tokio::spawn_blocking` and `.await` on it.